src/main/scala/cognite/spark/v1/RawSchemaInferrer.scala

```diff
@@ -0,0 +1,188 @@
+      case (JsonInteger, JsonDouble) => true
+      case (_, JsonString) => true
+      case (JsonObject(x), JsonObject(y)) =>
+        x.forall {
+          case (k, v) =>
+            y.get(k).exists(isAssignableTo(v, _))
+        }
+      case (JsonArray(x), JsonArray(y)) => isAssignableTo(x, y)
+      case (_, _) if from == to => true
+      case _ => false
+    }
+
+  def unify(a: JsonType, b: JsonType): JsonType =
+    (a, b) match {
+      case (x: JsonObject, y: JsonObject) => unifyObjects(x, y)
+      case (_, _) if isAssignableTo(b, a) => a
+      case (_, _) if isAssignableTo(a, b) => b
+      case (JsonArray(x), JsonArray(y)) => JsonArray(unify(x, y))
+
+      // anything can be put into a string - we just put the serialized JSON into it
+      case (_, _) => JsonString
+    }
+
+  def unifyObjects(a: JsonObject, b: JsonObject): JsonObject = {
+    if (isAssignableTo(b, a)) {
+      a
+    } else if (isAssignableTo(a, b)) {
+      b
+    } else {
+      val keys = a.fields.keySet.union(b.fields.keySet)
+      val fields = keys.iterator.map(k => k -> unify(a.field(k), b.field(k))).toMap
+      JsonObject(fields)
+    }
+  }
+
+  def toSparkSchema(j: JsonType): DataType = {
+    import org.apache.spark.sql.types._
+    j match {
+      case JsonNull => StringType
+      case JsonInteger => LongType
+      case JsonDouble => DoubleType
+      case JsonString => StringType
+      case JsonBool => BooleanType
+      case obj: JsonObject => toSparkSchema(obj)
+      case JsonArray(element) => ArrayType(toSparkSchema(element))
+    }
+  }
+
+  def toSparkSchema(j: JsonObject): StructType =
+    StructType(j.fields.map {
+      case (name, t) => StructField(name, toSparkSchema(t), nullable = true)
+    }.toArray)
+
+
+  private def skipInteger(str: String, index: Int): Int = {
+    var i = index
+    while (i < str.length && str(i) >= '0' && str(i) <= '9') {
+      i += 1
+    }
+    i
+  }
+
+  private def skipPlusMinus(str: String, index: Int): Int =
+    if (str(index) == '-' || str(index) == '+') {
+      index + 1
+    } else {
+      index
+    }
+
+  private def isIntegerString(str: String): Boolean = {
+    val i = skipPlusMinus(str, 0)
+
+    if (i + 18 < str.length) {
+      // won't fit into int64
+      false
+    } else {
+      val i2 = skipInteger(str, i)
+      i2 > i && i2 == str.length
+    }
+  }
+
+  private def isFloatString(str: String): Boolean = {
+    if ("NaN".equals(str)) {
+      return true
+    }
+    var i = skipPlusMinus(str, 0)
+
+    if (str.length == i + "Infinity".length && str.endsWith("Infinity")) {
+      return true
+    }
+
+    i = skipInteger(str, i)
+    if (i < str.length && str(i) == '.') {
+      i = skipInteger(str, i + 1)
+    }
+    if (i < str.length && (str(i) == 'e' || str(i) == 'E')) {
+      i = skipPlusMinus(str, i + 1)
+      val i2 = skipInteger(str, i)
+      if (i2 == i) {
+        return false
+      }
+      i = i2
+    }
+
+    i == str.length
+  }
```

Comment:
I don't get the logic here 😂 , this might need some comments to explain I think, I have a suggestion below to avoid using return
```scala
if ("NaN".equals(str)) {
  true
} else {
  var i = skipPlusMinus(str, 0)
  var whatever = true
  if (str.length == i + "Infinity".length && str.endsWith("Infinity")) {
    true
  } else {
    i = skipInteger(str, i)
    if (i < str.length && str(i) == '.') {
      i = skipInteger(str, i + 1)
    }
    if (i < str.length && (str(i) == 'e' || str(i) == 'E')) {
      i = skipPlusMinus(str, i + 1)
      val i2 = skipInteger(str, i)
      if (i2 == i) {
        whatever = false
      }
      i = i2
    }
    i == str.length && whatever
  }
}
```

Reply:
I assume it's just copied from `JsonInferSchema.infer`?

Reply:
No it isn't, I'll add a few comments. (Spark uses jackson parser which seems to do this somehow automatically).

Reply:
I have tried to better explain what the code does. I first implemented it using a regex, but that was pretty slow, so I just handwrote the parser.

General principle is that I scan the entire input string with the `i` index and check if all characters are allowed in this context. The parsed format is something like this: `123[.456][e-10]`, so:
* skip the numbers
* if there is decimal part, skip the `.` and skip another sequence of numbers
* if there is the exp part, skip `e` and the number
* if we made it to the end (i==str.length), it looks like a valid number

Reply:
Can't we just do it simple:
```scala
def isFloat(s: String) =  
  try {
    s.toFloat
    true
  } catch {
    case _ : Throwable => false
  }
```
It works even for `NaN`, `+-Infinity` and `123[.456][e-10] `

Reply:
`def isFloat(s: String) = Try(s.toFloat).isSuccess` if we want to do it that way. ;) Maybe less efficient, but a lot easier.

Maybe we should use circe?
Their decoder https://github.com/circe/circe/blob/main/modules/core/shared/src/main/scala/io/circe/Decoder.scala#L675 uses https://github.com/circe/circe/blob/main/modules/core/shared/src/main/scala/io/circe/JsonNumber.scala#L248 which uses https://github.com/circe/circe/blob/e6685317bc9e6e65539bb6194900e0e754fbcb7c/modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala#L257

Reply:
Ok, I'll try that, but I expect to get much worse results from that. This function is called for every string value in the RAW table and most of the time it will fail - for 350 string columns * only 20k rows -> 7 mil exceptions. I don't think JVM can throw this many exceptions in few seconds

Reply:
Yeah, it might be a bad choice for performance reasons, but the circe option might not be. It just seems to me like there might be quite a few edge cases that we haven't thought about. I also looked at jsoniter-scala, and that code looks even more scary, but also due to performance optimizations.

Reply:
Ok no.... I'm stupid, seems that Spark does not do this inferrence from strings, so we can just remove it :)